File: src/FSharpVSPowerTools/ShellEventListener.cs
using Microsoft.VisualStudio;
using Microsoft.VisualStudio.Shell.Interop;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FSharpVSPowerTools
{
    /// <summary>
    /// A class that encapsulates listening for shell events.
    /// </summary>
    public class ShellEventListener : IVsBroadcastMessageEvents, IDisposable
    {
        private const uint WM_SYSCOLORCHANGE = 0x0015;

        private IVsShell shellService;
        private uint broadcastEventCookie;

        private readonly IServiceProvider serviceProvider;

        public event EventHandler OnThemeChanged;

        public ShellEventListener(IServiceProvider serviceProvider)
        {
            this.serviceProvider = serviceProvider;
        }

        private void AdviseBroadcastMessages()
        {
            shellService = serviceProvider.GetService(typeof(SVsShell)) as IVsShell;

            if (shellService != null)
                ErrorHandler.ThrowOnFailure(shellService.AdviseBroadcastMessages(this, out broadcastEventCookie));
        }

        private void UnadviseBroadcastMessages()
        {
            if (shellService != null && broadcastEventCookie != 0)
            {
                shellService.UnadviseBroadcastMessages(broadcastEventCookie);
                broadcastEventCookie = 0;
            }
        }

        /// <summary>
        /// Watch for system color and font change event from Visual Studio.
        /// </summary>
        public int OnBroadcastMessage(uint msg, IntPtr wParam, IntPtr lParam)
        {
            if (msg == WM_SYSCOLORCHANGE)
            {
                if (OnThemeChanged != null)
                {
                    OnThemeChanged(this, EventArgs.Empty);
                }
            }
            return VSConstants.S_OK;
        }

        public void Initialize()
        {
            AdviseBroadcastMessages();
        }

        public void Dispose()
        {
            UnadviseBroadcastMessages();
        }
    }
}

File: src/FSharpVSPowerTools.Logic/ProjectSystem/ShellEventListener.fs
namespace FSharpVSPowerTools.ProjectSystem

open FSharpVSPowerTools
open Microsoft.VisualStudio.Shell.Interop
open Microsoft.VisualStudio
open System
open System.ComponentModel.Composition
open Microsoft.VisualStudio.Shell

/// Encapsulate listening for shell events.
[<Export>]
type ShellEventListener () as self =
    let shellService = Package.GetService<SVsShell, IVsShell>() |> Option.ofNull
    let broadcastEventCookie = ref 0u
    do shellService 
       |> Option.iter (fun shell ->
            ErrorHandler.ThrowOnFailure(shell.AdviseBroadcastMessages(self, broadcastEventCookie)) |> ignore)
    let themeChanged = DelegateEvent<EventHandler>()

    static let [<Literal>] WM_SYSCOLORCHANGE = 0x0015u

    [<CLIEvent>]
    member __.ThemeChanged = themeChanged.Publish

    interface IVsBroadcastMessageEvents with
        member __.OnBroadcastMessage(msg: uint32, _wParam: nativeint, _lParam: nativeint): int = 
            if msg = WM_SYSCOLORCHANGE then
                themeChanged.Trigger([|box self; box EventArgs.Empty|])
            VSConstants.S_OK
        
    interface IDisposable with
        member __.Dispose() = 
            shellService 
            |> Option.iter (fun shell -> 
                 shell.UnadviseBroadcastMessages(!broadcastEventCookie) |> ignore
                 broadcastEventCookie := 0u)
 